ddtrace/internal/_encoding.pyx

```diff
@@ -0,0 +1,357 @@
+                L = len(o)
+                if L > ITEM_LIMIT:
+                    PyErr_Format(ValueError, b"%.200s object is too large", Py_TYPE(o).tp_name)
+                rawval = o
+                ret = msgpack_pack_bin(&self.pk, L)
+                if ret == 0:
+                    ret = msgpack_pack_raw_body(&self.pk, rawval, L)
+            elif PyUnicode_CheckExact(o):  #  if strict_types else PyUnicode_Check(o):
+                if self.encoding == NULL:
+                    ret = msgpack_pack_unicode(&self.pk, o, ITEM_LIMIT)
+                    if ret == -2:
+                        raise ValueError("unicode string is too large")
+                else:
+                    o = PyUnicode_AsEncodedString(o, self.encoding, self.unicode_errors)
+                    L = len(o)
+                    if L > ITEM_LIMIT:
+                        raise ValueError("unicode string is too large")
+                    ret = msgpack_pack_raw(&self.pk, L)
+                    if ret == 0:
+                        rawval = o
+                        ret = msgpack_pack_raw_body(&self.pk, rawval, L)
+            elif PyDict_CheckExact(o):
+                d = <dict>o
+                L = len(d)
+                if L > ITEM_LIMIT:
+                    raise ValueError("dict is too large")
+                ret = msgpack_pack_map(&self.pk, L)
+                if ret == 0:
+                    for k, v in d.items():
+                       ret = self._pack(k)
+                       if ret != 0: break
+                       ret = self._pack(v)
+                       if ret != 0: break
+            elif PyList_CheckExact(o):
+                # Expect a list of traces or a list of spans
+
+                L = len(o)
+                if L > ITEM_LIMIT:
+                    raise ValueError("list is too large")
+
+                ret = msgpack_pack_array(&self.pk, L)
+                if ret != 0:
+                    break
+
+                if L > 0 and PyList_CheckExact(o[0]):
+                    # List of traces
+                    for i in range(L):
+                        span = o[i]
+                        ret = self._pack(span)
+                        if ret != 0: break
+                else:
+                    # List of spans
+                    for i in range(L):
+                        ret = self._pack(o[i])
+                        if ret != 0: break
+
+            elif type(o) is Span:
+                L = 12
+                if L > ITEM_LIMIT:
+                    raise ValueError("list is too large")
+
+                has_span_type = <bint>(o.span_type is not None)
+                has_meta = <bint>(len(o.meta) > 0)
+                has_metrics = <bint>(len(o.metrics) > 0)
+
+                L = L - (1 - has_span_type) - (1 - has_meta) - (1 - has_metrics)
+
+                ret = msgpack_pack_map(&self.pk, L)
+
+                if ret == 0:
+                    ret = self._pack_bytes(<char *>b"trace_id")
```

Comment:
wonder if we gain anything from inlining `_pack_bytes` here? `cdef` should be fast

Reply:
We can probably achieve this with an inline function. I'll try converting _pack_bytes to be an inline function instead of a method.

Reply:
do you think we'd actually gain anything at all?

if we 100% inline, then we are skipping a function call, but probably won't gain enough there, but if we inline to skip over some type/bounds checks that aren't necessary, then maybe (but maybe we can afford to remove any checks)

Reply:
Tried it and we do get some gains. Hard to quantify exactly how much, but there's a tangible improvement it seems:

## python 3:
### before

```
--------------------------------------------------------------------------------------- benchmark 'encoding': 3 tests ---------------------------------------------------------------------------------------
Name (time in ms)                             Min                 Max                Mean            StdDev              Median               IQR            Outliers       OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_1000_span_trace_custom         5.2127 (1.0)        7.2161 (1.0)        5.8392 (1.0)      0.3259 (1.0)        5.7838 (1.0)      0.4562 (1.0)          45;4  171.2577 (1.0)         171           1
test_encode_1000_span_trace                8.5683 (1.64)      28.1719 (3.90)       9.4243 (1.61)     1.9590 (6.01)       9.1895 (1.59)     0.4978 (1.09)          1;1  106.1091 (0.62)         96           1
test_encode_1000_span_trace_fallback     198.6543 (38.11)    201.2765 (27.89)    199.7139 (34.20)    0.9717 (2.98)     199.4761 (34.49)    1.0242 (2.25)          2;0    5.0072 (0.03)          5           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------- benchmark 'encoding.small': 2 tests -----------------------------------------------------------------------------------
Name (time in us)                       Min                 Max                Mean            StdDev              Median               IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_trace_small_custom      64.7078 (1.0)       86.4936 (1.0)       69.4768 (1.0)      3.7133 (1.0)       68.7429 (1.0)      3.6980 (1.0)          28;8       14.3933 (1.0)         142         100
test_encode_trace_small            126.0124 (1.95)     188.3710 (2.18)     135.4189 (1.95)     8.0008 (2.15)     134.8413 (1.96)     7.0486 (1.91)         10;2        7.3845 (0.51)         77         100
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------- benchmark 'encoding.small.multi': 2 tests ------------------------------------------------------------------------------
Name (time in ms)                           Min                Max              Mean            StdDev            Median               IQR            Outliers       OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_trace_small_multi_custom     3.1451 (1.0)       4.9664 (1.0)      3.4786 (1.0)      0.2554 (1.0)      3.4099 (1.0)      0.2284 (1.0)          19;7  287.4750 (1.0)         136           2
test_encode_trace_small_multi            6.4301 (2.04)     21.3525 (4.30)     7.2237 (2.08)     1.8430 (7.22)     6.9058 (2.03)     0.4966 (2.17)          3;6  138.4330 (0.48)        123           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
### after
```
--------------------------------------------------------------------------------------- benchmark 'encoding': 3 tests ---------------------------------------------------------------------------------------
Name (time in ms)                             Min                 Max                Mean            StdDev              Median               IQR            Outliers       OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_1000_span_trace_custom         5.3183 (1.0)        7.2103 (1.0)        5.7886 (1.0)      0.3077 (1.0)        5.7217 (1.0)      0.3569 (1.0)          45;7  172.7535 (1.0)         174           1
test_encode_1000_span_trace                8.6917 (1.63)      25.4237 (3.53)       9.4444 (1.63)     1.6387 (5.33)       9.2111 (1.61)     0.3965 (1.11)          1;5  105.8830 (0.61)        101           1
test_encode_1000_span_trace_fallback     198.4867 (37.32)    203.9461 (28.29)    200.6900 (34.67)    2.3365 (7.59)     200.2065 (34.99)    3.9596 (11.10)         1;0    4.9828 (0.03)          5           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------- benchmark 'encoding.small': 2 tests -----------------------------------------------------------------------------------
Name (time in us)                       Min                 Max                Mean            StdDev              Median               IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_trace_small_custom      61.2236 (1.0)       91.6967 (1.0)       67.0073 (1.0)      4.2223 (1.0)       66.1946 (1.0)      4.4449 (1.0)          36;7       14.9237 (1.0)         155         100
test_encode_trace_small            126.9934 (2.07)     151.7563 (1.65)     136.2673 (2.03)     4.9573 (1.17)     135.7910 (2.05)     6.9569 (1.57)         23;1        7.3385 (0.49)         74         100
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------- benchmark 'encoding.small.multi': 2 tests ------------------------------------------------------------------------------
Name (time in ms)                           Min                Max              Mean            StdDev            Median               IQR            Outliers       OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_trace_small_multi_custom     3.0215 (1.0)       3.8439 (1.0)      3.2450 (1.0)      0.1525 (1.0)      3.2097 (1.0)      0.1702 (1.0)          43;8  308.1651 (1.0)         152           2
test_encode_trace_small_multi            6.4966 (2.15)     20.5880 (5.36)     7.2128 (2.22)     1.6907 (11.08)    6.9705 (2.17)     0.3813 (2.24)          2;5  138.6415 (0.45)        131           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

## python 2

### before
```
--------------------------------------------------------------------------------------- benchmark 'encoding': 3 tests ----------------------------------------------------------------------------------------
Name (time in ms)                             Min                 Max                Mean             StdDev              Median                IQR            Outliers      OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_1000_span_trace_custom        10.5500 (1.0)      120.4650 (4.73)      20.3133 (1.39)     17.0410 (7.48)      12.0060 (1.0)      19.6059 (13.14)        10;1  49.2287 (0.72)         69           1
test_encode_1000_span_trace               12.8379 (1.22)      25.4540 (1.0)       14.5672 (1.0)       2.2769 (1.0)       13.9611 (1.16)      1.4917 (1.0)           6;6  68.6475 (1.0)          65           1
test_encode_1000_span_trace_fallback     250.5860 (23.75)    266.9230 (10.49)    256.6416 (17.62)     6.3324 (2.78)     254.2419 (21.18)     7.5364 (5.05)          1;0   3.8965 (0.06)          5           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

------------------------------------------------------------------------------------ benchmark 'encoding.small': 2 tests ------------------------------------------------------------------------------------
Name (time in us)                       Min                 Max                Mean             StdDev              Median                IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_trace_small_custom      88.9993 (1.0)      156.5194 (1.0)      102.0038 (1.0)      11.8719 (1.16)      99.0891 (1.0)       8.5920 (1.0)         14;11        9.8036 (1.0)         103         100
test_encode_trace_small            142.7794 (1.60)     194.2897 (1.24)     156.6929 (1.54)     10.2694 (1.0)      155.1247 (1.57)     13.7651 (1.60)         17;2        6.3819 (0.65)         64         100
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------- benchmark 'encoding.small.multi': 2 tests ------------------------------------------------------------------------------
Name (time in ms)                           Min                Max              Mean            StdDev            Median               IQR            Outliers       OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_trace_small_multi_custom     4.5309 (1.0)       9.0189 (1.0)      5.6420 (1.0)      1.1031 (1.0)      5.0950 (1.0)      1.4337 (2.22)         25;3  177.2423 (1.0)         127           1
test_encode_trace_small_multi            7.7090 (1.70)     18.2650 (2.03)     8.6866 (1.54)     1.4113 (1.28)     8.3139 (1.63)     0.6466 (1.0)          6;11  115.1201 (0.65)        107           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```


### after

```
-------------------------------------------------------------------------------------- benchmark 'encoding': 3 tests ---------------------------------------------------------------------------------------
Name (time in ms)                             Min                 Max                Mean            StdDev              Median               IQR            Outliers      OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_1000_span_trace_custom        10.0381 (1.0)       11.9171 (1.0)       10.6940 (1.0)      0.4311 (1.0)       10.6690 (1.0)      0.5805 (1.0)          32;1  93.5107 (1.0)          88           1
test_encode_1000_span_trace               12.3880 (1.23)      24.1070 (2.02)      13.4087 (1.25)     1.3836 (3.21)      13.2790 (1.24)     0.6154 (1.06)          1;1  74.5787 (0.80)         67           1
test_encode_1000_span_trace_fallback     248.5950 (24.77)    257.2050 (21.58)    252.5710 (23.62)    3.4132 (7.92)     251.3280 (23.56)    5.0768 (8.74)          2;0   3.9593 (0.04)          5           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------- benchmark 'encoding.small': 2 tests -----------------------------------------------------------------------------------
Name (time in us)                       Min                 Max                Mean            StdDev              Median               IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_trace_small_custom      84.1188 (1.0)      106.7901 (1.0)       90.3827 (1.0)      4.2066 (1.0)       89.6406 (1.0)      6.1440 (1.15)         39;1       11.0641 (1.0)         113         100
test_encode_trace_small            143.0488 (1.70)     170.3596 (1.60)     151.2291 (1.67)     4.9071 (1.17)     150.6805 (1.68)     5.3620 (1.0)          18;2        6.6125 (0.60)         67         100
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------- benchmark 'encoding.small.multi': 2 tests ------------------------------------------------------------------------------
Name (time in ms)                           Min                Max              Mean            StdDev            Median               IQR            Outliers       OPS            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_encode_trace_small_multi_custom     4.0990 (1.0)       5.0285 (1.0)      4.4449 (1.0)      0.2092 (1.0)      4.3970 (1.0)      0.2648 (1.0)          32;3  224.9764 (1.0)         107           2
test_encode_trace_small_multi            7.5500 (1.84)     16.2640 (3.23)     8.3875 (1.89)     1.1393 (5.45)     8.1871 (1.86)     0.5309 (2.00)          2;2  119.2248 (0.53)        107           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```